pyramid_swagger/load_schema.py

```diff
@@ -160,49 +139,85 @@ def get_model_resolver(schema):
+    objects. Each field is a :class:`SchemaValidator`.
     """
     __slots__ = ()
 
-
-def build_request_to_schemas_map(schema):
-    """Take the swagger schema and build a map from incoming path to a
-    jsonschema for requests and responses."""
-    request_to_schema = {}
+    @classmethod
+    def from_operation(cls, operation, models, resolver):
+        args = []
+        for schema, validator in [
+            (build_param_schema(operation, 'query'), Draft3Validator),
+            (build_param_schema(operation, 'path'), Draft3Validator),
+            (build_param_schema(operation, 'header'), Draft3Validator),
+            (extract_body_schema(operation, models), Draft4Validator),
+            (extract_response_body_schema(operation, models),
+                Draft4Validator),
+        ]:
+            args.append(SchemaValidator.from_schema(
+                schema,
+                resolver,
+                validator))
+
+        return cls(*args)
+
+
+class SchemaValidator(object):
+
+    def __init__(self, schema, validator):
+        self.schema = schema
+        self.validator = validator
+
+    @classmethod
+    def from_schema(cls, schema, resolver, validator_class):
+        return cls(
+            schema,
+            validator_class(schema, resolver=resolver, types=EXTENDED_TYPES))
+
+    def validate(self, values):
+        if not self.schema:
+            return
+        self.validator.validate(values)
+
+
+def build_request_to_validator_map(schema, resolver):
+    """Build a mapping from :class:`RequestMatcher` to :class:`ValidatorMap`
+    for each operation in the API spec. This mapping may be used to retrieve
+    the appropriate validators for a request.
+    """
     schema_models = schema.get('models', {})
-    for api in schema['apis']:
-        path = api['path']
-        for operation in api['operations']:
-            # Now that we have the necessary info for this particular
-            # path/method combination, build our dict.
-            key = (path, operation['method'])
-            request_to_schema[key] = SchemaMap(
-                request_query_schema=extract_query_param_schema(operation),
-                request_path_schema=extract_path_schema(operation),
-                request_body_schema=extract_body_schema(
-                    operation,
-                    schema_models
-                ),
-                response_body_schema=extract_response_body_schema(
-                    operation,
-                    schema_models
-                ),
-            )
-
-    return request_to_schema
+    return dict(
+        (
+            RequestMatcher(api['path'], operation['method']),
+            ValidatorMap.from_operation(operation, schema_models, resolver)
+        )
+        for api in schema['apis']
+        for operation in api['operations']
+    )
+
+
+class RequestMatcher(object):
+    """Match a :class:`pyramid.request.Request` to a swagger Operation"""
```

Comment:
Let's encode (document? assert?) assumptions about the request already being routed (probably better than doing that routing here)